src/Git.hs

```diff
@@ -472,120 +468,115 @@ runGit userConfig repoDir operation =
 
-    LastTag sha cont -> do
+    LastTag sha -> do
       result <- callGitInRepo ["describe", "--abbrev=0", "--tags", refSpec sha]
-      pure $ cont $ either (const Nothing) Just result
+      pure $ either (const Nothing) Just result
 
-    ShortLog shaStart shaEnd cont -> do
+    ShortLog shaStart shaEnd -> do
       result <- callGitInRepo ["shortlog", refSpec shaStart <> ".." <> refSpec shaEnd]
       case result of
         Left (_, message) -> do
           logWarnN $ "git shortlog failed. Reason: " <> message
-          pure $ cont Nothing
-        Right changelog -> do
-          pure $ cont $ Just changelog
+          pure Nothing
+        Right changelog -> pure $ Just changelog
 
-    Tag sha t (TagMessage m) cont -> do
+    Tag sha t (TagMessage m) -> do
       result <- callGitInRepo ["tag", "-a", refSpec t, "-m", Text.unpack m, refSpec sha]
       case result of
         Left (_, message) -> do
           logWarnN $ "git tag failed. Reason: " <> message
-          pure $ cont $ TagFailed message
+          pure $ TagFailed message
         Right _ -> do
           logInfoN $ format "tagged {} with {}" [show sha, show t]
-          pure $ cont $ TagOk t
+          pure $ TagOk t
 
-    DeleteBranch branch cont -> cont <$ callGitInRepo ["branch", "-d", refSpec branch]
+    DeleteBranch branch -> void $ callGitInRepo ["branch", "-d", refSpec branch]
 
-    DeleteTag t cont -> cont <$ callGitInRepo ["tag", "-d", refSpec t]
+    DeleteTag t -> void $ callGitInRepo ["tag", "-d", refSpec t]
 
-    CheckOrphanFixups sha branch cont -> do
+    CheckOrphanFixups sha branch -> do
       result <- let branch' = refSpec branch
                     sha' = refSpec sha
                 in callGitInRepo ["log", Text.unpack $ format "{}..{}" [branch',sha'], "--format=%s"]
       case result of
         Left (_, message) -> do
           logWarnN $ "git log failed. Reason: {}" <> message
-          pure $ cont False
+          pure False
         Right logResponse -> do
           let anyOrphanFixups = any (\x -> "fixup!" `Text.isPrefixOf` x) $ Text.lines logResponse
           when anyOrphanFixups $
             logWarnN "there is one ore more fixup commits not belonging to any other commit"
-          pure $ cont anyOrphanFixups
+          pure anyOrphanFixups
 
 -- Interpreter that runs only Git operations that have no side effects on the
 -- remote; it does not push.
 runGitReadOnly
-  :: forall m a
-   . MonadIO m
-  => MonadLogger m
+  :: forall es a
+  .  (IOE :> es, MonadLoggerEffect :> es)
   => UserConfiguration
   -> FilePath
-  -> GitOperationFree a
-  -> m a
-runGitReadOnly userConfig repoDir operation =
-  let
-    unsafeResult = runGit userConfig repoDir operation
-  in
-    case operation of
+  -> Eff (GitOperation : es) a
+  -> Eff es a
+runGitReadOnly userConfig repoDir = runGit userConfig repoDir . augmentedGitOperation
+  where
+    augmentedGitOperation = interpose $ \_ operation -> case operation of
```

Comment:
This `runGitReadOnly` function is an interesting one, it takes the regular `runGit` function, but _overrides_ some of the actions, making them readonly. With `effectful`, the way to do that is with [`interpose`](https://hackage.haskell.org/package/effectful-core-2.2.2.2/docs/Effectful-Dispatch-Dynamic.html#v:interpose).